File: packages/control-contract/src/work.ts
import type { Revision } from "./revision.js";
import type { TypedId } from "./ids.js";

/** Global control-plane kinds; local Work cards must not invent more kinds. */
export type WorkItemKind = "task" | "ticket" | "campaign";

/** Human-facing global Work lifecycle. */
export type WorkStatus =
  | "backlog"
  | "todo"
  | "in_progress"
  | "in_review"
  | "done"
  | "blocked"
  | "cancelled";

/** Execution lifecycle, deliberately separate from the Work lifecycle. */
export type ExecutionPhase =
  | "none"
  | "queued"
  | "planning"
  | "awaiting_plan_approval"
  | "running"
  | "awaiting_input"
  | "awaiting_approval"
  | "verifying"
  | "reviewing"
  | "retrying"
  | "paused"
  | "failed"
  | "needs_attention"
  | "terminal";

/** Canonical global Work projection; parentage is not a dependency edge. */
export type ControlWorkItem = {
  readonly id: TypedId;
  readonly project_id: TypedId;
  readonly goal_id: TypedId | null;
  readonly parent_work_item_id: TypedId | null;
  readonly kind: WorkItemKind;
  readonly title: string;
  readonly description: string;
  readonly status: WorkStatus;
  readonly execution_phase: ExecutionPhase;
  readonly revision: Revision;
  readonly created_at: string;
  readonly updated_at: string;
};

File: src-tauri/crates/altai-control-protocol/src/work.rs
//! Canonical global WorkItem contract.
//!
//! This models global planning authority only. The workspace-local ledger
//! retains exact run/session facts and must synchronize through explicit
//! leases and revisions rather than replacing these fields.

use crate::{GoalId, ProjectId, Revision, WorkItemId};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum WorkItemKind {
    Task,
    Ticket,
    Campaign,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum WorkStatus {
    Backlog,
    Todo,
    InProgress,
    InReview,
    Done,
    Blocked,
    Cancelled,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ExecutionPhase {
    None,
    Queued,
    Planning,
    AwaitingPlanApproval,
    Running,
    AwaitingInput,
    AwaitingApproval,
    Verifying,
    Reviewing,
    Retrying,
    Paused,
    Failed,
    NeedsAttention,
    Terminal,
}

/// The control-plane projection of a work object. `parent_work_item_id` is
/// decomposition only; dependencies are modeled separately before dispatch.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct WorkItem {
    pub id: WorkItemId,
    pub project_id: ProjectId,
    pub goal_id: Option<GoalId>,
    pub parent_work_item_id: Option<WorkItemId>,
    pub kind: WorkItemKind,
    pub title: String,
    pub description: String,
    pub status: WorkStatus,
    pub execution_phase: ExecutionPhase,
    pub revision: Revision,
    pub created_at: String,
    pub updated_at: String,
}
